### PR TITLE
Adds versions to Js and Stylesheet within modules

### LINKED ISSUE #24410: Ability to add versions to js and css files
Hello guys,
#### Describe the bug
when we change css and js files for a module, we need to add a version to update clients caches to get up to dated files.
There is a need to specify the version when we register js and css files with these methods in controllers: 

- **registerJavascript**
- **registerStylesheet**

#### Steps to reproduce

1. Fresh install
2. Make sure the CCC is disabled
3. Enable Debug mode
4. Change anything in CSS/JS
5. Changes aren't visible on front for client

#### Additional information

* PrestaShop version: 1.7.7.x
* PHP version: 7.3

### PR DESCRIPTION
**Pull request by samberrry**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 1.7.7.x
| Description?      | This PR adds the ability to add versions to js and css files in modules. In some cases module changes will not be applied due to client's cache and needs to add a version to js/css files. 
| Type?             | new feature
| Category?         | FO 
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | #24410
| How to test?      | Add "version" property to params in registerStylesheet and registerJavascript methods 
| Possible impacts? | nothing

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24409)

### CODE REVIEW

**Comment by Progi1984** [comment]
@samberrry Could you validate the issue before the PR, please ?

**Comment by kpodemski** [comment]
There's no need for this change I think, you can provide version as query string in URL to your asset.

**Comment by matks** [comment]
> There's no need for this change I think, you can provide version as query string in URL to your asset.

@kpodemski you mean doing `src="/js/a/b/c/picker.js?v=7278"` ?

**Comment by kpodemski** [comment]
@matks 

```
 $this->registerStylesheet(
            'module-style',
            'modules/'.$this->module->name.'/views/css/instruction.css?v=123'
        );
```

**Comment by samberrry** [comment]
@Progi1984 ok I try to do that.

**Comment by samberrry** [comment]
@kpodemski before coding this PR I tried your solution. It does not work.

File: classes/assets/JavascriptManager.php [hunk 1]
```diff
@@ -73,12 +74,15 @@ public function unregisterById($idToRemove)
...
     }
 
-    protected function add($id, $fullPath, $position, $priority, $inline, $attribute, $server)
+    protected function add($id, $fullPath, $position, $priority, $inline, $attribute, $server, $version)
     {
         $priority = is_int($priority) ? $priority : self::DEFAULT_PRIORITY;
         $position = $this->getSanitizedPosition($position);
         $attribute = $this->getSanitizedAttribute($attribute);
 
+        //../javascript.js?{version}
```
Comment (kpodemski) [inline classes/assets/JavascriptManager.php]:
```suggestion
```

File: classes/assets/JavascriptManager.php [hunk 2]
```diff
@@ -73,12 +74,15 @@ public function unregisterById($idToRemove)
...
+        $fullPath = ($version)? $fullPath.'?'.$version : $fullPath;
```
Comment (kpodemski) [inline classes/assets/JavascriptManager.php]:
```suggestion
        $fullPath = $version ? $fullPath.'?'.$version : $fullPath;
```

make sure that if someone is using`//domain.com?stylesheet=1` using `external` type, we still can use our own versioning

File: classes/assets/StylesheetManager.php [hunk 3]
```diff
@@ -87,11 +88,14 @@ public function getList()
         return $this->list;
     }
 
-    protected function add($id, $fullPath, $media, $priority, $inline, $server)
+    protected function add($id, $fullPath, $media, $priority, $inline, $server, $version)
     {
         $priority = is_int($priority) ? $priority : self::DEFAULT_PRIORITY;
         $media = $this->getSanitizedMedia($media);
 
+        //../stylesheet.css?{version}
```
Comment (kpodemski) [inline classes/assets/StylesheetManager.php]:
```suggestion
```

File: classes/assets/StylesheetManager.php [hunk 4]
```diff
@@ -87,11 +88,14 @@ public function getList()
...
+        $fullPath = ($version)? $fullPath.'?'.$version : $fullPath;
```
Comment (kpodemski) [inline classes/assets/StylesheetManager.php]: same as above, code formatting + handling external resources

File: classes/controller/FrontController.php [hunk 5]
```diff
@@ -1083,7 +1084,7 @@ public function registerStylesheet($id, $relativePath, $params = [])
                 . ($this->stylesheetManager->getFullPath($relativePath) ?? $relativePath);
             $params['server'] = 'remote';
         }
-        $this->stylesheetManager->register($id, $relativePath, $params['media'], $params['priority'], $params['inline'], $params['server']);
+        $this->stylesheetManager->register($id, $relativePath, $params['media'], $params['priority'], $params['inline'], $params['server'], true, $params['version']);
```
Comment (kpodemski) [inline classes/controller/FrontController.php]: please add `needRtl` to default parameters and use it 👍

**Review by kpodemski — CHANGES_REQUESTED** [review]

**Comment by kpodemski** [comment]
> @kpodemski before coding this PR I tried your solution. It does not work.

you're right, `?v=X` doesn't work with new methods for registering assets

**Comment by samberrry** [comment]
@kpodemski thank you

File: classes/assets/JavascriptManager.php (hunk 2 above)
Comment (samberrry) [reply classes/assets/JavascriptManager.php]: @kpodemski yes I've tested it with `external` type. I uploaded changes on:
https://ps1773.binshops.com/
as you see the following link works fine:
https://ps1773.binshops.com/modules/ps_imageslider/css/homeslider.css?v1.2

File: classes/controller/FrontController.php (hunk 5 above)
Comment (samberrry) [reply classes/controller/FrontController.php]: For sure.

**Comment by kpodemski** [comment]
Hey @samberrry - could you rebase your PR to `develop`? it's not possible to release it in 1.7.7 or 1.7.8

File: classes/controller/FrontController.php [hunk 6]
```diff
@@ -1084,7 +1084,8 @@ public function registerStylesheet($id, $relativePath, $params = [])
                 . ($this->stylesheetManager->getFullPath($relativePath) ?? $relativePath);
             $params['server'] = 'remote';
         }
-        $this->stylesheetManager->register($id, $relativePath, $params['media'], $params['priority'], $params['inline'], $params['server'], true, $params['version']);
+
+        $this->stylesheetManager->register($id, $relativePath, $params['media'], $params['priority'], $params['inline'], $params['server'], isset($params['needRtl']) ? $params['needRtl'] : true, $params['version']);
```
Comment (kpodemski) [inline classes/controller/FrontController.php]: @samberrry maybe add `needRtl` to `$default_params`?

File: classes/assets/JavascriptManager.php [hunk 7]
```diff
@@ -51,12 +51,13 @@ public function register(
         $priority = self::DEFAULT_PRIORITY,
         $inline = false,
         $attribute = null,
-        $server = 'local'
+        $server = 'local',
+        $version = null
```
Comment (Progi1984) [inline classes/assets/JavascriptManager.php]:
```suggestion
        ?string $version = null
```

File: classes/assets/JavascriptManager.php [hunk 8]
```diff
@@ -73,12 +74,14 @@ public function unregisterById($idToRemove)
         }
     }
 
-    protected function add($id, $fullPath, $position, $priority, $inline, $attribute, $server)
+    protected function add($id, $fullPath, $position, $priority, $inline, $attribute, $server, $version)
```
Comment (Progi1984) [inline classes/assets/JavascriptManager.php]:
```suggestion
    protected function add($id, $fullPath, $position, $priority, $inline, $attribute, $server, ?string $version)
```

File: classes/assets/StylesheetManager.php [hunk 9]
```diff
@@ -53,18 +53,19 @@ public function register(
         $priority = self::DEFAULT_PRIORITY,
         $inline = false,
         $server = 'local',
-        $needRtl = true
+        $needRtl = true,
+        $version = null
```
Comment (Progi1984) [inline classes/assets/StylesheetManager.php]:
```suggestion
        string $version = null
```

File: classes/assets/StylesheetManager.php [hunk 10]
```diff
@@ -87,11 +88,13 @@ public function getList()
         return $this->list;
     }
 
-    protected function add($id, $fullPath, $media, $priority, $inline, $server)
+    protected function add($id, $fullPath, $media, $priority, $inline, $server, $version)
```
Comment (Progi1984) [inline classes/assets/StylesheetManager.php]:
```suggestion
    protected function add($id, $fullPath, $media, $priority, $inline, $server, ?string $version)
```

**Review by Progi1984 — COMMENTED** [review]
Some strict typing, please

File: classes/controller/FrontController.php (hunk 6 above)
Comment (samberrry) [reply classes/controller/FrontController.php]: done in new PR

File: classes/assets/JavascriptManager.php (hunk 7 above)
Comment (samberrry) [reply classes/assets/JavascriptManager.php]: done in new PR

**Comment by samberrry** [comment]
new PR:
https://github.com/PrestaShop/PrestaShop/pull/24656

**Pull request closed without merging**